client/components/board.jsx

```diff
@@ -36,10 +38,19 @@ class Board extends React.Component {
     }
     const { row, col } = this.state.selected;
     const input = this.state.challenge;
-    input[row][col] = event.target.value;
+    this.state.previousMove.push({ challenge: JSON.parse(JSON.stringify(input)), selected: JSON.parse(JSON.stringify(this.state.selected)) });
+    input[row][col] = parseInt(event.target.value);
     this.setState({ challenge: input, selected: null });
   }
 
+  handleUndo() {
+    if (!this.state.previousMove.length) {
+      return;
+    }
+    const { challenge, selected } = this.state.previousMove.pop();
+    this.setState({ challenge, selected });
```

Comment:
We do not pop() from state arrays. always copy the array first before changing elements.